README.md
[![CI Status](https://github.com/sealvault/sealvault/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/sealvault/sealvault/actions/workflows/ci.yml) 

```text
███████╗███████╗ █████╗ ██╗    ██╗   ██╗ █████╗ ██╗   ██╗██╗  ████████╗
██╔════╝██╔════╝██╔══██╗██║    ██║   ██║██╔══██╗██║   ██║██║  ╚══██╔══╝
███████╗█████╗  ███████║██║    ██║   ██║███████║██║   ██║██║     ██║   
╚════██║██╔══╝  ██╔══██║██║    ╚██╗ ██╔╝██╔══██║██║   ██║██║     ██║   
███████║███████╗██║  ██║███████╗╚████╔╝ ██║  ██║╚██████╔╝███████╗██║   
╚══════╝╚══════╝╚═╝  ╚═╝╚══════╝ ╚═══╝  ╚═╝  ╚═╝ ╚═════╝ ╚══════╝╚═╝
```

SealVault is an open source project to build a self-custody Web3 key manager app
that is primarily used for everyday browsing with social and gaming dapps. 

Learn more on the [website](https://sealvault.org) or subscribe to our
[newsletter.](https://sealvault.substack.com/subscribe)

## [Developer Docs](https://sealvault.org/dev-docs/)

Documentation for SealVault developers containing design documents on the
[security](https://sealvault.org/dev-docs/design/security-model/) and [privacy](https://sealvault.org/dev-docs/design/privacy-model/) models of the application and more.

## Components

### [Core](./core)

Cross-platform key and transaction management for Web3 written in Rust. 
Uses SQLite for storage and depends on platform-specific keychain to store 
key encryption keys.
Exposed to host languages via [uniffi-rs](https://github.com/mozilla/uniffi-rs).

### [iOS](./ios)

iOS Web3 key manager app with a built-in web browser. Written in Swift with 
the SwiftUI framework.

### [Tools](./tools)

Contains various development tools:

- [Development server](./tools/dev-server) for integration testing in-page providers
- [Fuzzer](./tools/fuzz)
- [TLA+ models](./tools/tla)

### [Website](./website)

Provides landing page and features for end users, developer documentation for
SealVault developers and the SealVault blog. Powered by [Material for
MkDocs](https://squidfunk.github.io/mkdocs-material/).

## Development

On first time setup, install the [requirements](#development-requirements)
listed below and then run `./ci.py` from the repo root to check if everything
was installed correctly. (:warning: The CI script will reset all iOS simulators
on your machine.)

The CI script will also create the generated files for you that are a
prerequisites for an Xcode build. Alternatively you can generate these files by
running `fastlane pre_build` in the `ios` directory.

### Development Requirements

- MacOS Monterey and an Apple Silicon Mac for iOS development. You may be able 
  to work on an Intel Mac, but we won't be able to help you with Rust-related 
  build issues as the maintainers are on Apple Silicon.
- XCode 14 IDE and CLI tools get the Xcode IDE from the app store and then run
  `xcode-select --install` to install CLI tools.
- The [Homebrew](https://brew.sh/) macOS package manager, then run `brew bundle`
  from the repo root to install the dependencies specified in the
  [`Brewfile`.](./Brewfile)
- Run `rustup-init` which was installed from the [`Brewfile`](./Brewfile) 
  toolchain and follow the instructions to set up your Rust toolchain.
- The [Diesel](https://diesel.rs/) ORM CLI that we use to manage Sqlite (it's
  very important to use the bundled Sqlite version as outdated Sqlite make cause
  migrations to fail):
  - `cargo install diesel_cli --no-default-features --features sqlite-bundled` 
- The [Solidity](https://docs.soliditylang.org/en/v0.8.16/index.html) compiler 
  version manager and `solc 0.8.16` that are needed to compile test contracts 
  for unit tests:
  - `cargo install svm-rs` then `svm install 0.8.16`
- The [Foundry](https://book.getfoundry.sh/getting-started/installation) 
  Ethereum development tools of which Anvil is needed for unit tests.
- Get Ankr account to get a free [Advanced
  API](https://www.ankr.com/rpc/advanced-api) endpoint and then set up
  [development secrets](#development-secrets).

### Development Secrets

Navigate to the [Ankr Advanced API
console](https://www.ankr.com/rpc/advanced-api), copy the endpoint that looks
like this: https://rpc.ankr.com/multichain/1234abcde. Extract the API key which
is the last part of the URL.

Once you have the API key create the `ios/Config.xcconfig` file that is ignored
by git with the following contents:

```shell
ANKR_API_KEY=1234abcde
```

### Tests

Run all tests from the repo root with: `./ci.py`. (:warning: The CI script will
reset all iOS simulators on your machine.)

### Useful commands

- Update Rust toolchain to latest nightly: `rustup update`.
  - We need nightly for iOS simulator on MacOS Apple Silicon.
  - If the CI fails for you locally, make sure you're on the latest nightly.
- Clean Rust build from repo root: `cargo clean`.
- Rebuild Rust XCFramework for iOS and regenerate Swift bindings from `ios`
  directory: `fastlane pre_build`
- Clean Xcode build from `ios` directory: `fastlane clean`.


CHANGELOG.md
# Changelog

## 2023-06-11

### iOS Beta v0.9.0

- Add ZkSync support.
- Disable dapp allotment transfers when new dapp key is created as we don't know
  which chain will be used when the key is created.
- Make sheet titles consistent.
- Fetch common tokes by default with native token API.
- Allow users to add custom tokens to wallet.
- Make transfer form pop out instead of being behind a navigation link.
- Fix tokens not loading from Ankr API.

## 2023-04-09

### iOS Beta v0.8.0

- Increase transaction reliability by using gas oracles to estimate gas fees and
  gradually increasing gas price for unconfirmed transactions.
- Add support for `wallet_watchAsset` MetaMask RPC method to add custom fungible
  tokens to wallet.
- Add support Filecoin (FEVM) Hyperspace testnet. There is no NFT support and no
  automatic custom token discovery support yet (tracking issues:
  [#112](https://github.com/sealvault/sealvault/issues/112) and
  [#113](https://github.com/sealvault/sealvault/issues/113)).

## 2023-04-02

### iOS Beta v0.7.2

- Update logo
- Improve app startup time ([#109](https://github.com/sealvault/sealvault/pull/109]))
- Support opening block explorer for the appropriate chain if an address is used
  on multiple chains.

## 2023-03-16

### iOS Beta v0.7.1

- Inject Ethereum provider only into main frame by default
- Refactor signing key to minimize time in memory ([#20](https://github.com/sealvault/sealvault/issues/20))
- Under-the-hood improvements for upcoming cross-connect feature

#### Improvements

- Display NFTs for address
- Auto-detect new chains ([#105](https://github.com/sealvault/sealvault/issues/105))
- Under-the-hood improvements for upcoming cross-connect feature

## 2023-02-15

### iOS Beta v0.7.0

#### Improvements

- Display NFTs for address
- Auto-detect new chains ([#105](https://github.com/sealvault/sealvault/issues/105))
- Under-the-hood improvements for upcoming cross-connect feature

#### Fixes

- Fix missing notification if transfer to profile wallet fails

## 2023-01-17

### iOS Beta v0.6.2

#### Improvements

- Add success notification on enabling backups

#### Fixes

- Fix notification resetting selection.
  ([#100](https://github.com/sealvault/sealvault/issues/100))


## 2023-01-11

### iOS Beta v0.6.1

#### Improvements

- Add active profile checkmark in profile view

#### Fixes

- Make launching dapp from inactive profile turn that profile active.
  ([#99](https://github.com/sealvault/sealvault/issues/99))

### iOS Beta v0.6.0

#### Features

- [Self-custody backups](https://sealvault.org/backup)
- Create new profiles and allow switching profiles

#### Improvements

- Change chain to network in UI
- Improve address menu UX
- Add profile name to wallet row
- Remove labels for nav bar buttons
- New profile pic selection with seals
- Change default tab to profiles
- Allow copying profile wallet with swipe
- Add browser history. Previously, user was always browsing in incognito mode.
  Now the browser history is retained until the user switches profiles.

#### Fixes

- Exclude DB file from iCloud Backup since it won't work due to missing keys
  from local keychain.

## 2022-12-23

### iOS Beta v0.5.3

#### Improvements

- Improve default dapps on new installs
- Revert to old logo

## 2022-12-19

### iOS Beta v0.5.2

#### Improvements

- Improve browser address bar UI
- Improve top dapps UI
- Improve dark mode colors

## 2022-12-13

### iOS Beta v0.5.0

#### Improvements

- Change accounts to profiles
- Improve token display
- Improve add dapp modal text

## 2022-12-07

### iOS Beta v0.4.4

#### Improvements

- Allow dismissing banners by swiping up

#### Fixes

- Fix not auto-hiding banners
- Update mint NFT domain to the one that it newly redirects

## 2022-11-28

### iOS Beta v0.4.3

#### Improvements

- Change default dapps to [quickswap.exchange](https://quickswap.exchange/) and [mintnft.today](https://mintnft.today/)

## 2022-11-22

### iOS Beta v0.4.2

#### Improvements

- Increase token transfer amount input field width.
- Increase the number of times we show the default token amount transfer
  disclosure to new users to 5. After they've added 5 new dapps, the disclosure
  is hidden behind an expand button.

#### Fixes

- Fix opening dapp in browser with long press not loading page.

#### Fixes

## 2022-11-18

### iOS Beta v0.4.1

#### Fixes

- Fix address bar not showing due to too many top dapps
- Allow dismissing address bar keyboard by scrolling page down 

### iOS Beta v0.4.0

#### Improvements

- Add top dapps list as default view in browser windows
- Add two dapps by default for new users to help with onboarding
- Streamline transfer form
- Change nav bar title to more legible

## 2022-11-15

## iOS Beta v0.3.3

#### Improvements

- Add chain name to dapp transaction error message on which the error happened.

#### Fixes

- Fix error message display for "insufficient funds for gas * price + value"
  message.

## 2022-11-13

### iOS Beta v0.3.2

#### Improvements

- Change nav tab bar labels to "Browser Tab 1", "Accounts", "Browser Tab 2"
  based on beta feedback.

## 2022-11-09

### iOS Beta v0.3.1

#### Fixes

- Change default home page in in-app web browser due to App Store Review
  requirements.

## 2022-11-01

### iOS Beta v0.3.0

#### Features

- Allow changing connected chain for Ethereum dapps from app. This is important,
because some dapps don't support changing the chain from their UI and just error
out if the user isn't on the expected chain.

#### Fixes

- Fix can't connect [crypto coven](https://cryptocoven.xyz/)

## 2022-10-30

### iOS Beta v0.2.2

#### Improvements

- Add popup banners for feedback about blockchain interactions. This gives users
  more info and lets them switch views while waiting for feedback. Events:
  - Token transfer submitted to mempool success/error.
  - Token transfer confirmed success/error.
  - Dapp allotment transfer success/error.

core/src/utils.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use std::{collections::HashSet, path::Path, time::SystemTime};

use chacha20poly1305::aead::generic_array::{ArrayLength, GenericArray};
use chrono::{DateTime, FixedOffset, SecondsFormat, Utc};
use email_address::EmailAddress;
use lazy_static::lazy_static;
use rand::{thread_rng, RngCore};
use tempfile::NamedTempFile;
use url::Url;
use uuid::Uuid;

use crate::Error;

/// Create an RFC339 timestamp, eg.: "2018-01-26T18:30:09.453Z".
/// From: https://stackoverflow.com/a/64148017
pub fn rfc3339_timestamp() -> String {
    let now = SystemTime::now();
    let dt: DateTime<Utc> = now.into();
    dt.to_rfc3339_opts(SecondsFormat::Millis, true)
}

pub fn parse_rfc3339_timestamp(s: &str) -> Result<DateTime<FixedOffset>, Error> {
    DateTime::parse_from_rfc3339(s).map_err(|err| Error::Retriable {
        error: err.to_string(),
    })
}

/// Number of seconds since unix epoch
pub fn unix_timestamp() -> i64 {
    let dt: DateTime<Utc> = SystemTime::now().into();
    dt.timestamp()
}

pub fn path_to_string(path: &Path) -> Result<String, Error> {
    path.to_str()
        .map(|s| s.to_string())
        .ok_or_else(|| Error::Fatal {
            error: "Got invalid utf-8 string for path".into(),
        })
}

pub fn tmp_file() -> Result<NamedTempFile, Error> {
    NamedTempFile::new().map_err(|err| Error::Retriable {
        error: format!("Failed to create temporary file with error: '{err}'"),
    })
}

/// Generate a v4 UUID.
pub fn new_uuid() -> String {
    Uuid::new_v4().to_string()
}

// Minimum secure length for random bytes (128 bit).
const MIN_RANDOM_BYTES_LENGTH: usize = 16;

pub fn try_fill_random_bytes(buffer: &mut [u8]) -> Result<(), Error> {
    // This is most likely a logic error and potentially dangerous. It can happen for example that
    // the buffer has 0 length when passing a vector created with a capacity, but not filled yet.
    if buffer.len() < MIN_RANDOM_BYTES_LENGTH {
        return Err(Error::Fatal {
            error: "Insecure secret buffer length.".into(),
        });
    }
    let mut rng = thread_rng();
    rng.try_fill_bytes(buffer).map_err(|err| Error::Fatal {
        error: format!(
            "Failed to generate random bytes with error code: {:?}",
            err.code()
        ),
    })?;
    Ok(())
}

/// Generate cryptographically secure random bytes for a generic array.
/// Fails if the OS random generator fails.
pub fn try_random_bytes<N: ArrayLength<u8>>() -> Result<GenericArray<u8, N>, Error> {
    let mut buffer: GenericArray<u8, N> = GenericArray::default();
    try_fill_random_bytes(&mut buffer)?;
    Ok(buffer)
}

/// Generate cryptographically secure random bytes for a generic array.
/// Fails if the OS random generator fails.
pub fn blake3_hash<T: AsRef<[u8]>>(value: T) -> blake3::Hash {
    let mut hasher = blake3::Hasher::new();
    hasher.update(value.as_ref());
    hasher.finalize()
}

lazy_static! {
    static ref ALLOWED_SCHEMES: HashSet<&'static str> =
        ["http", "https", "file", "ftp",].into();
}

/// Turn a possible URI that a user typed into a browser address bar into a fully qualified URL.
/// Returns the URL if it's valid, else None.
/// Defaults to http if no scheme is specified to cover IP addresses as https upgrades are handled
/// elsewhere.
pub fn uri_fixup(input: impl AsRef<str>) -> Option<String> {
    let trimmed = input.as_ref().trim();

    // Quoted strings and email addresses should be passed on to search engine.
    if trimmed.contains('"') || EmailAddress::is_valid(trimmed) {
        return None;
    }

    let url = Url::parse(trimmed).or_else(|_| Url::parse(&format!("http://{}", trimmed)));

    url.ok().and_then(|u| {
        if !ALLOWED_SCHEMES.contains(u.scheme()) {
            return None;
        }

        if let Some(domain) = u.domain() {
            // If the domain has only one part and it's not localhost, we assume that it won't
            // resolve. This might cause problems in some development setups, or on networks that
            // make use of partially qualified domain names, but we cannot do better without doing
            // a DNS lookup.
            if domain != "localhost" && domain_part_count(domain) <= 1 {
                return None;
            }
        }

        let mut s = u.to_string();
        if s.ends_with('/') {
            s.pop();
        }
        Some(s)
    })
}

fn domain_part_count(domain: &str) -> usize {
    domain.split('.').filter(|s| !s.is_empty()).count()
}

#[cfg(test)]
mod tests {
    use anyhow::Result;
    use chacha20poly1305::aead::generic_array::typenum::U32;

    use super::*;

    #[test]
    fn rfc_timestamp() -> Result<()> {
        let timestamp = rfc3339_timestamp();
        let _ = parse_rfc3339_timestamp(&timestamp)?;
        Ok(())
    }

    #[test]
    fn fills_random_bytes() -> Result<()> {
        const SIZE: usize = 32;
        let zeroes: Vec<u8> = vec![0; SIZE];

        let mut arr: Vec<u8> = vec![0; SIZE];
        try_fill_random_bytes(&mut arr)?;

        assert_ne!(arr, zeroes);
        assert_eq!(arr.len(), zeroes.len());

        Ok(())
    }

    #[test]
    fn random_bytes_is_not_default() -> Result<()> {
        type Array = GenericArray<u8, U32>;
        let default = Array::default();

        let rand_bytes: Array = try_random_bytes()?;

        assert_ne!(rand_bytes, default);
        assert_eq!(rand_bytes.len(), default.len());

        Ok(())
    }

    #[test]
    fn uri_fixup_cases() {
        fn ss(s: &str) -> Option<String> {
            Some(s.into())
        }

        // Test cases from: https://github.com/brave/brave-ios/blob/c93e42d0996039492a7b6b3783dad12574405dcb/Tests/ClientTests/SearchTests.swift

        // Check valid URLs. We can load these after some fixup.
        assert_eq!(
            uri_fixup("http://www.mozilla.org"),
            ss("http://www.mozilla.org")
        );
        // assert_eq!(url_fixup("about:"), ss("about:"));
        // assert_eq!(url_fixup("about:config"), ss("about:config"));
        // assert_eq!(url_fixup("about: config"), ss("about:%20config"));
        assert_eq!(uri_fixup("file:///f/o/o"), ss("file:///f/o/o"));
        assert_eq!(
            uri_fixup("ftp://ftp.mozilla.org"),
            ss("ftp://ftp.mozilla.org")
        );
        assert_eq!(uri_fixup("foo.bar"), ss("http://foo.bar"));
        assert_eq!(uri_fixup(" foo.bar "), ss("http://foo.bar"));
        // TODO failing case (strips port 80)
        // assert_eq!(uri_fixup("[::1]:80"), ss("http://[::1]:80"));
        assert_eq!(
            uri_fixup("[2a04:4e42:400::288]"),
            ss("http://[2a04:4e42:400::288]")
        );
        // TODO failing case (strips port 80)
        // assert_eq!(uri_fixup("[2a04:4e42:600::288]:80"), ss("http://[2a04:4e42:600::288]:80"));
        assert_eq!(
            uri_fixup("[2605:2700:0:3::4713:93e3]:443"),
            ss("http://[2605:2700:0:3::4713:93e3]:443")
        );
        // TODO failing cases (compresses to hex)
        // assert_eq!(uri_fixup("[::192.9.5.5]"), ss("http://[::192.9.5.5]"));
        // assert_eq!(uri_fixup("[::192.9.5.5]:80"), ss("http://[::192.9.5.5]:80"));
        // assert_eq!(uri_fixup("[::192.9.5.5]/png"), ss("http://[::192.9.5.5]/png"));
        // assert_eq!(uri_fixup("[::192.9.5.5]:80/png"), ss("http://[::192.9.5.5]:80/png"));
        assert_eq!(uri_fixup("192.168.2.1"), ss("http://192.168.2.1"));
        assert_eq!(uri_fixup("brave.io"), ss("http://brave.io"));
        assert_eq!(uri_fixup("brave.new.world"), ss("http://brave.new.world"));
        assert_eq!(
            uri_fixup("brave.new.world.test"),
            ss("http://brave.new.world.test")
        );
        assert_eq!(
            uri_fixup("brave.new.world.test.io"),
            ss("http://brave.new.world.test.io")
        );
        assert_eq!(
            uri_fixup("brave.new.world.test.whatever.io"),
            ss("http://brave.new.world.test.whatever.io")
        );
        // TODO failing case (converts to octets)
        // assert_eq!(uri_fixup("http://2130706433:8000/"), ss("http://2130706433:8000/"));
        assert_eq!(
            uri_fixup("http://127.0.0.1:8080"),
            ss("http://127.0.0.1:8080")
        );
        // TODO failing cases (expands to all octets)
        // assert_eq!(uri_fixup("http://127.0.1"), ss("http://127.0.1"));
        // assert_eq!(uri_fixup("http://127.1"), ss("http://127.1"));
        // assert_eq!(uri_fixup("http://127.1:8000"), ss("http://127.1:8000"));
        // assert_eq!(uri_fixup("http://1.1:80"), ss("http://1.1:80"));

        // Check invalid URLs. These are passed along to the default search engine.
        assert_eq!(uri_fixup("foobar"), None);
        assert_eq!(uri_fixup("foo bar"), None);
        assert_eq!(uri_fixup("mozilla. org"), None);
        // TODO failing case (parsed as IP address)
        // assert_eq!(uri_fixup("123"), None);
        assert_eq!(uri_fixup("a/b"), None);
        assert_eq!(uri_fixup("创业咖啡"), None);
        assert_eq!(uri_fixup("创业咖啡 中国"), None);
        assert_eq!(uri_fixup("创业咖啡. 中国"), None);
        assert_eq!(
            uri_fixup("data:text/html;base64,SGVsbG8gV29ybGQhCg=="),
            None
        );
        assert_eq!(
            uri_fixup("data://https://www.example.com,fake example.com"),
            None
        );
        // TODO failing cases (parsed as IP address)
        // assert_eq!(uri_fixup("1.2.3"), None);
        // assert_eq!(uri_fixup("1.1"), None);
        // assert_eq!(uri_fixup("127.1"), None);
        // assert_eq!(uri_fixup("127.1.1"), None);

        // Check invalid quoted URLs, emails, and quoted domains.
        // These are passed along to the default search engine.
        assert_eq!(uri_fixup(r#""123"#), None);
        assert_eq!(uri_fixup(r#""123""#), None);
        assert_eq!(uri_fixup(r#""ftp.mozilla.org"#), None);
        assert_eq!(uri_fixup(r#""ftp.mozilla.org""#), None);
        assert_eq!(uri_fixup(r#"https://"ftp.mozilla.org""#), None);
        assert_eq!(uri_fixup(r#"https://"ftp.mozilla.org"#), None);
        assert_eq!(uri_fixup("foo@brave.com"), None);
        assert_eq!(uri_fixup("\"foo@brave.com"), None);
        assert_eq!(uri_fixup("\"foo@brave.com\""), None);
        assert_eq!(uri_fixup(r#""创业咖啡.中国"#), None);
        assert_eq!(uri_fixup(r#""创业咖啡.中国""#), None);
        assert_eq!(uri_fixup("foo:5000"), None);
        assert_eq!(uri_fixup("http://::192.9.5.5"), None);
        assert_eq!(uri_fixup("http://::192.9.5.5:8080"), None);

        // Added for SealVault
        assert_eq!(uri_fixup("localhost"), ss("http://localhost"));
        assert_eq!(uri_fixup("http://localhost"), ss("http://localhost"));
        // TODO failing cases (returned with scheme)
        // assert_eq!(uri_fixup("localhost:8080/open-new-tab.html"), ss("http://localhost:8080/open-new-tab.html"));
        // assert_eq!(uri_fixup("localhost:1234"), ss("http://localhost:1234"));
        assert_eq!(
            uri_fixup("http://localhost:8080"),
            ss("http://localhost:8080")
        );
        assert_eq!(uri_fixup("https://example.com"), ss("https://example.com"));

        assert_eq!(uri_fixup("uniswap"), None);
    }
}
